src/checkers/inference/DefaultSlotManager.java

```diff
@@ -359,6 +370,23 @@ public ExistentialVariableSlot createExistentialVariableSlot(VariableSlot potent
         return existentialVariableSlot;
     }
 
+    @Override
+    public ArithmeticVariableSlot createArithmeticVariableSlot(AnnotationLocation location) {
```

Comment:
If I have source code:
```java
class A {
   @1 Object o1;
   @2 Object o2;

   void test1() {
       o1 + o2;
   }
   void test2() {
       o1 + o2;
   }
}
```
There would be two `ArithmeticVariableSlot` created, because two `o1+ o2` have different `AnnotationLocation`, correct? Then the cache seems not working anymore.

Reply:
Good point. Well the cache works if the intention is to cache on location. But it would be a unique second location that isn't necessary.

I'll switch back to caching on the 2 component slots + operation kind.

Reply:
I did a bunch of debugging and it turns out that when VariableAnnotator.handleBinaryTree() is called, the two operands have VariableSlots with different IDs, and the IDs are also different in each of the two test methods. Example debug output:

```
Created ArithmeticVariableSlot
  14 = VariableSlot(12) + VariableSlot(13)
  20 = VariableSlot(18) + VariableSlot(19)
```

Slot 14 is created for test1()'s addition, and Slot 20 is created for test2()'s addition.

If instead of @1 and @2 in the declaration, we had two constant slots for units annotations, we still get VariableSlots here during handleBinaryTree() and not ConstantSlots. I don't know why, but it seems substitution of constant slots is done at a later stage in the inference process.

So it seems to be impossible to create unique slots by caching on the operands slots and the operation, since the VarSlot IDs are different.

I'm going to keep it simple and cache on annotation location. Either way we are creating two ArithmeticVarSlots and thus two ArithmeticConstraints. There's no good way to make it even more efficient at this point.

Also, I think in real code we shouldn't see the same equation appear twice if proper procedural design is applied, ie we should have
```
void test2() {
   test1();
}

Reply:
I think we should find out the reason why two constant slots are not used instead VariableSlots with different IDs in different locations. Even though in both cases(caching on location and caching on operand + operation) create different `ArithmeticVarSlot`, it's the problem in VariableAnnotator(maybe other classes). Using location to cache will completely remove the possibility of optimization.

> Also, I think in real code we shouldn't see the same equation appear twice if proper procedural design is applied

We can never assume this. If you run real projects, there will be tons of scnarios that shocks you. Your solution is only for this testcase.

Reply:
Using location to cache is equivalent to not caching. As constraints are only generated once on each location. We'll end up with a cache that always grows bigger, but never hits cached results, I think.

Reply:
> I think we should find out the reason why two constant slots are not used instead VariableSlots with different IDs in different locations. 

Separate PR? I'm sure this problem affects not just arithmetic var slots.

> Using location to cache is equivalent to not caching. As constraints are only generated once on each location.

The constraint manager does not cache constraints, but instead, we have the hashcodes of each constraint used to filter duplicates during insertion to the constraint set. In the detailed implementation of almost all of the constraints, we use the component slots as part of the hashcode. None of them use the location.

Its also not universally true that constraints are generated only once per location. Anytime we get to QualHierarchy.isSubtype we are generating subtype constraints, and we call this a lot during the multiple passes of InferenceATF subclasses and tree annotators (eg within LUB). We actually generate many duplicate constraints but which get filtered out during insertion to the constraint set.

Reply:
> Using location to cache will completely remove the possibility of optimization.

What kind of optimization are you referring to?